packages/sol-compiler/src/compiler.ts

```diff
@@ -107,97 +150,100 @@ export class Compiler {
+        // batch input contracts together based on the version of the compiler that they require.
+        const versionToInputs: VersionToInputs = {};
+
+        // map contract paths to data about them for later verification and persistence
+        const contractPathToData: ContractPathToData = {};
+
+        for (const contractName of contractNames) {
+            const contractSource = this._resolver.resolve(contractName);
+            contractPathToData[contractSource.path] = {
+                contractName,
+                currentArtifactIfExists: await getContractArtifactIfExistsAsync(this._artifactsDir, contractName),
+                sourceTreeHashHex: `0x${this._getSourceTreeHash(
+                    path.join(this._contractsDir, contractSource.path),
+                ).toString('hex')}`,
+            };
+            if (!this._shouldCompile(contractPathToData[contractSource.path])) {
+                continue;
             }
-            solcjs = await response.text();
-            fs.writeFileSync(compilerBinFilename, solcjs);
+            const solcVersion = _.isUndefined(this._solcVersionIfExists)
+                ? semver.maxSatisfying(_.keys(binPaths), parseSolidityVersionRange(contractSource.source))
+                : this._solcVersionIfExists;
+            if (_.isUndefined(versionToInputs[solcVersion])) {
+                // no inputs batched yet for this version.
+                // prepare object to hold this batch.
+                versionToInputs[solcVersion] = {
+                    standardInput: {
+                        language: 'Solidity',
+                        sources: {},
+                        settings: this._compilerSettings,
+                    },
+                    contractsToCompile: [],
+                };
+            }
+            // add input to the right version batch
+            versionToInputs[solcVersion].standardInput.sources[contractSource.path] = {
+                content: contractSource.source,
+            };
+            versionToInputs[solcVersion].contractsToCompile.push(contractSource.path);
         }
-        const solcInstance = solc.setupMethods(requireFromString(solcjs, compilerBinFilename));
 
-        logUtils.log(`Compiling ${contractName} with Solidity v${solcVersion}...`);
-        const standardInput: solc.StandardInput = {
-            language: 'Solidity',
-            sources: {
-                [contractSource.path]: {
-                    content: contractSource.source,
-                },
-            },
-            settings: this._compilerSettings,
-        };
-        const compiled: solc.StandardOutput = JSON.parse(
-            solcInstance.compileStandardWrapper(JSON.stringify(standardInput), importPath => {
-                const sourceCodeIfExists = this._resolver.resolve(importPath);
-                return { contents: sourceCodeIfExists.source };
-            }),
-        );
+        for (const solcVersion of _.keys(versionToInputs)) {
```

Comment:
I know this is my suggestion. Feel this would be slightly more readable:

```
const solcVersions = _.keys(versionToInputs);
for (const solcVersion of solcVersions) {
```